include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,873 @@
+	 *  indicated by dict_len. if set to 0 and if support_dict ==
+	 *  ODP_SUPPORT_YES, then implementation will use dictionary length
+	 *  less than or equal to user input length in odp_comp_set_dict()
+	 *  and update used dictionary length at output of the call.
+	 *
+	 */
+	uint32_t dict_len;
+
+	/** Maximum compression level supported by implementation of this algo.
+	 * Indicates number of compression levels supported by implementation,
+	 *
+	 * where,
+	 *
+	 * 1 means fastest compression i.e. output produced at
+	 * best possible speed at the expense of compression quality, and
+	 *
+	 * max_level means best compression i.e.output produced is best possible
+	 * compressed content at the expense of speed.
+	 *
+	 * Example, if max_level = 4 , it means algorithm supports four levels
+	 * of compression from value 1 up to 4. User can set this value from
+	 * 1 (fastest compression) to 4 (best compression).
+	 * See RFC1950 for an example explanation to level.
+	 *
+	 * Value 0 mean implementation use its default value.
+	 *
+	 */
+	uint32_t max_level;
+
+	/* Supported hash algorithms */
+	odp_comp_hash_algos_t hash_algo;
+} odp_comp_alg_capability_t;
+
+/**
+ * Comp API dictionary type
+ * Consists of pointer to byte buffer. length of dictionary
+ * indicated by length parameter.
+ */
+typedef struct odp_comp_dict_t {
+	/** pointer to byte array */
+	uint8_t *buf;
+	/** length of the dictionary. */
+	uint32_t len;
+} odp_comp_dict_t;
+
+/**
+ * Comp API algorithm specific parameters
+ *
+ */
+typedef struct odp_comp_algo_param_t {
+	/** struct for defining deflate algorithm parameters.
+	* Also initialized by other deflate based algorithms , ex. ZLIB
+	*/
+	struct comp_alg_def_param {
+		/** compression level where
+		 * ODP_COMP_LEVEL_MIN <= level <= ODP_COMP_LEVEL_MAX
+		 */
+		odp_comp_level_t level;
+		/** huffman code to use */
+		odp_comp_huffman_code_t comp_code;
+	} deflate;
+
+	/** struct for defining zlib algorithm parameters.
+	 */
+	struct comp_alg_zlib_param {
+			/** deflate algo params */
+			struct comp_alg_def_param def;
+	} zlib;
+} odp_comp_alg_param_t;
+
+ /**
+ * Comp API session creation parameters
+ *
+ */
+typedef struct odp_comp_session_param_t {
+	/** Compress vs. Decompress operation */
+	odp_comp_op_t op;
+
+	/** Sync vs Async mode
+	 *
+	 * When mode = ODP_COMP_SYNC, odp_comp_xxx()
+	 * should be called.
+	 *
+	 * When mode = ODP_COMP_ASYNC, odp_comp_xxx_enq()
+	 * should be called.
+	 *
+	 * Use odp_comp_capability() for supported mode.
+	 *
+	 */
+	odp_comp_op_mode_t mode;
+
+	/** Compression algorithm
+	 *
+	 *  Use odp_comp_capability() for supported algorithms.
+	 */
+	odp_comp_alg_t comp_algo;
+
+	/** Hash algorithm
+	 *
+	 *  Use odp_comp_alg_capability() for supported hash algo for
+	 *  compression algo given as comp_alg. Implementation should not
+	 *  support hash only operation on data. output should always contain
+	 *  data + hash.
+	 *
+	 */
+	odp_comp_hash_alg_t hash_algo;
```

Comment:
Typo: "... given as comp_alg" -> "... given as comp_algo"

There are two SHOULDs, but I think those are not needed:
1) comp_algo is always set. It should be e.g. ALG_NULL by default. We have defined that NULL does not have hash capability. It would be also possible to leave NULL + hash to be implementation defined capability. When we say that ALG_NULL is only for test, it should be enough to discourage people from misusing this API for hashing.
2) What this means: "output should always contain data + hash" ? If we enable hashing, hash is always calculated. It's not optional for implementation to calculate it - right.

Reply:
1.comp_algo is always set. It should be e.g. ALG_NULL by default. 
>> I am bit confused. Are you suggesting adding a description to "comp ALG_NULL+hash"? We agreed in past that " user must use crypto for standalone hashing and compression algo NULL is only for testing purpose" and current description on usage of hash and ALG_NULL abide that.  So, we can leave current description as is.
2. What this means: "output should always contain data + hash" ? If we enable hashing, hash is always calculated. It's not optional for implementation to calculate it - right.
>> I will rephrase it to "output *will* always contain data + hash". 
Please approve the change.

Reply:
Should is not must. So, "Implementation should not support hash only operation on data" does not specify anything concrete. Implementation can still offer hash with NULL.

I think we can leave it as a capability (we have the API for that), but say that NULL algo is for testing purposes only == not performance optimized etc.